source/core_mqtt_serializer.c

```diff
@@ -357,95 +588,1308 @@ static MQTTStatus_t checkPublishRemainingLength( size_t remainingLength,
+                encodedByte = pBuffer[ bytesDecoded ];
+                remainingLength += ( ( size_t ) encodedByte & 0x7FU ) * multiplier;
+                multiplier *= 128U;
+                bytesDecoded++;
+            }
+
+            /* If the response is incorrect then
+             * break out of the loop. */
+            if( remainingLength == MQTT_REMAINING_LENGTH_INVALID )
+            {
+                break;
+            }
+        } while( ( encodedByte & 0x80U ) != 0U );
+
+        if( status == MQTTSuccess )
+        {
+            /* Check that the decoded remaining length conforms to the MQTT specification. */
+            expectedSize = remainingLengthEncodedSize( remainingLength );
+
+            if( bytesDecoded != expectedSize )
+            {
+                LogError( ( "Expected and actual length of decoded bytes do not match.\n" ) );
+                status = MQTTBadResponse;
+            }
+            else
+            {
+                *pLength = remainingLength;
+            }
+        }
+
+        return status;
+    }
+
+    static MQTTStatus_t decodeuint32_t( uint32_t * pProperty,
+                                        size_t * pPropertyLength,
+                                        bool * pUsed,
+                                        const uint8_t ** pIndex )
+    {
+        const uint8_t * pVariableHeader = *pIndex;
+        MQTTStatus_t status = MQTTSuccess;
+
+        /*Protocol error to include the same property twice.*/
+        if( *pUsed == true )
+        {
+            status = MQTTProtocolError;
+        }
+        /*Validate the length and decode.*/
+        else if( *pPropertyLength < sizeof( uint32_t ) )
+        {
+            status = MQTTMalformedPacket;
+        }
+        else
+        {
+            *pProperty = UINT32_DECODE( pVariableHeader );
+            pVariableHeader = &pVariableHeader[ sizeof( uint32_t ) ];
+            *pUsed = true;
+            *pPropertyLength -= sizeof( uint32_t );
+        }
+
+        *pIndex = pVariableHeader;
+        return status;
+    }
+
+    static MQTTStatus_t decodeuint16_t( uint16_t * pProperty,
+                                        size_t * pPropertyLength,
+                                        bool * pUsed,
+                                        const uint8_t ** pIndex )
+    {
+        const uint8_t * pVariableHeader = *pIndex;
+        MQTTStatus_t status = MQTTSuccess;
+
+        /*Protocol error to include the same property twice.*/
+
+        if( *pUsed == true )
+        {
+            status = MQTTProtocolError;
+        }
+        /*Validate the length and decode.*/
+
+        else if( *pPropertyLength < sizeof( uint16_t ) )
+        {
+            status = MQTTMalformedPacket;
+        }
+        else
+        {
+            *pProperty = UINT16_DECODE( pVariableHeader );
+            pVariableHeader = &pVariableHeader[ sizeof( uint16_t ) ];
+            *pUsed = true;
+            *pPropertyLength -= sizeof( uint16_t );
```

Comment:
Will there be a scenario where pPropertyLength can be greater than sizeof( uint16_t ). Same is true for all similar cases in the other decode functions